cache.go

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) 2016 turbobytes
+//
+// This file is part of geoipdb, a library of GeoIP related helper functions
+// for TurboBytes stack.
+//
+// MIT License
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package geoipdb
+
+import (
+	"time"
+)
+
+// cacheTTL is the expiration time of a cache entry.
+const cacheTTL = time.Hour * 24
+
+// cacheEntry is the data we want to keep cached.
+type cacheEntry struct {
+	// ASN number
+	asn string
+	// ASN description
+	descr string
+	// Due date of this entry
+	due time.Time
+}
+
+// cache allows manipulating cached data.
+type cache struct {
```

Comment:
> So you have map `ip` storing IP-to-ASN with ASN names, and `asn` for storing ASN-to-IP to use for reverse looks.

Correct. I could do a brute force search in `ip` map for reverse finding IPs for a given ASN as well, but I felt maintaining a dedicated map for the reverse lookup would pay in the long run.

> Some feedback. 
> 
> 1. Suppose we send multiple queries belonging to same ASN but different IPs, those will all be cache MISS leading to accessing database and 3rd party sources.

Correct. Just reminding that the only "direct" lookup that is cache aware is `geoipdb.LookupAsn()`; all other queries (overrides, maxmind, ipinfo) remain uncached.

So in the case `LookupAsn()` is called several times with distinct IPs belonging to the same ASN, map `ip` will contain several keys (the IPs), and the elements of these keys will contain the same value for the `asn` field. The `descr` field may or may not be the same, according to the precedence rules used by `LookupAsn()`.

> 2. If maxmind says 1.1.1.1 is AS555 and ipinfo says it is AS666, then when we do a reverse search for 1.1.1.1 will it show descr for AS555 or AS666 ?

The reverse lookup method in geoipdb is `LookupIp(asn)`, returning all IPs from the cache that belongs to the given ASN. Therefore, if cache is empty, a reverse lookup for any one of these ASNs will return empty.

As currently implemented, the cache learns when `LookupAsn()` is called. The first call to `LookupAsn(1.1.1.1)` will cache miss, and the library resorts to the precedence rules for looking up 1.1.1.1. The first external resource to be queried is maxmind, returning AS555 and a description. Assuming there is no override for AS555, `LookupAsn(1.1.1.1)` returns (AS555, "maxmind description") with the side effect of caching these values. Further calls to `LookupAsn(1.1.1.1)` will cache hit, returning the same data as before -- at least until the 1 day expiration time is reached or `PurgeAsnCache()` is called.

In this situation:

- `LookupIp(AS555)` will return [1.1.1.1] -- and `LookupAsn(1.1.1.1)` can be issued as usual for getting the cached description ("maxmind description" in this case), completing the reverse lookup in two steps.
- `LookupIp(AS666)` will return an empty list, as there is no mention of AS666 in cache. Here `LookupAsn()` is useless. Clients can resort to `OverridesLookup()` or `CymruDnsLookup()` as fallbacks for getting a description for AS666, but the reverse lookup, as conceptually defined, fails because we didn't get any IPs that belong to this ASN.